### DISCUSSION TITLE
Context propagation across threads

### DISCUSSION THREAD

**Discussion by schelwa-tibco** [open-telemetry/opentelemetry-java#2884] [Q&A] [answered]
I am trying to trace request that is processed asynchronously. As request is received on a thread, some processing is done and the request is put on a queue. The queue is serviced by a executor service thread pool. To trace this request, I am trying to use Context propagation. As the request is received, the context is `injected` on the carries (i.e. eventContext).

		Context currentContext = Context.current();
                Span span = tracer.spanBuilder(spanName).setParent(currentContext).startSpan();
		Context context = currentContext.with(span);
		PROPAGATOR.inject(context, eventContext, SETTER);
		scope = context.makeCurrent();
		// request processing code here
		span.end();
		scope.close();

Then on the executor service thread, context is `extracted` from the carrier (i.e. eventContext).

		Context parentContext = PROPAGATOR.extract(Context.current(), eventContext, GETTER);
		Span span = tracer.spanBuilder(spanName).setParent(parentContext).startSpan();
		context = parentContext.with(span);
		scope = context.makeCurrent();
		// some more request processing code here
		span.end();
		scope.close();

Is my understanding correct? This is not working for me. The spans don't share the same traceId hence not traceable as one request on Jaeger.

**Comment by jkwatson** [comment, marked as answer]
This does look roughly correct. Without doing some detailed debugging, though, it's a little hard to see where things could be going wrong. Do you have a full example somewhere that we could take a look at?

**Reply by schelwa-tibco** [reply]
Does `inject` call adds traceId and other relevant information on carrier? I added some debug statements on the SETTER, but I dont it getting executed.

**Reply by jkwatson** [reply]
That depends entirely on the propagator that you're using, which should implement how to propagate the needed information.

**Reply by schelwa-tibco** [reply]
I am using `GlobalOpenTelemetry.getPropagators().getTextMapPropagator()`. Do I need implement custom propagator for such a use case?

**Reply by kenfinnigan** [reply]
Was a Propagator installed into the OpenTelemetry SDK? I was bitten with that last week because I'd forgotten to do that, so no injection/extraction was happening.

Something like:
```
        OpenTelemetrySdkBuilder builder = OpenTelemetrySdk.builder();
        builder.setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()));
        builder.buildAndRegisterGlobal();
```

**Reply by schelwa-tibco** [reply]
Thanks Ken! That's what was missing.

**Comment by Oberon00** [comment]
If you just want to propagate across threads within the same process, you don't need context propagators (extract/inject). You can just pass the `Context` object directly to another thread (it is immutable and thus thread-safe).

**Reply by schelwa-tibco** [reply]
I ended up doing this. But, wanted to find out whats missing with my use of context propagation.
